### PR TITLE
Update `userLists`, `web3listener`, and `settings` Redux files to use TypeScript

### PR DESCRIPTION
**Pull request by marco**
Here is some initial work on converting some files to TypeScript. I based most of the changes off of the patterns I saw in #2495.

As far as I can tell, these changes don't break any `yarn lint` rules, but it's definitely possible I'm missing something!

### CODE REVIEW

File: src/redux/settings.ts [hunk 1]
```diff
@@ -102,15 +155,15 @@ export const settingsChangeNativeCurrency = nativeCurrency => async dispatch =>
 };
 
 // -- Reducer --------------------------------------------------------------- //
-export const INITIAL_STATE = {
+export const INITIAL_STATE: SettingsState = {
   accountAddress: '',
   chainId: 1,
   language: 'en',
   nativeCurrency: 'USD',
   network: networkTypes.mainnet,
```
Comment (jinchung) [inline src/redux/settings.ts:163]: we can use the enum: `Network.mainnet` and then remove the import of `networkTypes`

File: src/redux/settings.ts [hunk 2]
```diff
@@ -25,7 +26,51 @@ const SETTINGS_UPDATE_NETWORK_SUCCESS =
...
-export const settingsLoadState = () => async dispatch => {
+
+/**
+ * The current `settings` state.
+ */
+interface SettingsState {
+  accountAddress: string;
+  chainId: number;
+  language: string;
+  nativeCurrency: string;
```
[the remark this answers is not in the document]
Comment (jinchung) [reply src/redux/settings.ts:37]: this is a great point! let's make a separate linear ticket for the `language` and `nativeCurrency` types so we can keep track of it without it blocking this PR!

File: src/redux/userLists.ts [hunk 3]
```diff
@@ -21,16 +23,90 @@ const USER_LISTS_UPDATE_LISTS = 'userLists/USER_LISTS_UPDATE_LISTS';
...
+export const userListsLoadState = () => async (
+  dispatch: AppDispatch,
+  getState: () => AppState
+) => {
   const { network } = getState().settings;
 
   dispatch({ type: USER_LISTS_LOAD_REQUEST });
   try {
-    const defaultLists = DefaultTokenLists[network] || [];
-    const userLists = await getUserLists(network);
```
[the remark this answers is not in the document]
Comment (jinchung) [reply src/redux/userLists.ts:31]: yes you're right - good catch!

File: src/redux/userLists.ts [hunk 4]
```diff
@@ -21,16 +23,90 @@ const USER_LISTS_UPDATE_LISTS = 'userLists/USER_LISTS_UPDATE_LISTS';
...
+  type: typeof USER_LISTS_READY;
+}
+
+interface UserListsClearStateAction {
+  type: typeof USER_LISTS_CLEAR_STATE;
+}
+
+export const userListsLoadState = () => async (
+  dispatch: AppDispatch,
+  getState: () => AppState
```
Comment (jinchung) [inline src/redux/userLists.ts:96]: You can import `AppGetState` and replace `() => AppState` with `AppGetState`

Comment (jinchung) [reply src/redux/userLists.ts:96]: Same with all the other places using `() => AppState`

File: src/redux/userLists.ts [hunk 5]
```diff
@@ -21,16 +23,90 @@ const USER_LISTS_UPDATE_LISTS = 'userLists/USER_LISTS_UPDATE_LISTS';
...
+  | UserListsClearStateAction;
+
+interface UserListsLoadRequestAction {
+  type: typeof USER_LISTS_LOAD_REQUEST;
+}
+
+interface UserListsLoadSuccessAction {
+  type: typeof USER_LISTS_LOAD_SUCCESS;
+  payload: {
+    lists: UserList[];
```
Comment (jinchung) [inline src/redux/userLists.ts:68]: this isn't part of your TS changes, but noticed we can improve this: the payload could be the `UserList[]` directly instead of being an object with `lists` inside of it. This also means updating line 115 and line 241 accordingly

**Comment by jinchung** [comment]
on line 125: 
```      dispatch({
        payload: true,
        type: USER_LISTS_READY,
      });```
the `payload: true` is not needed

File: src/redux/web3listener.ts [hunk 6]
```diff
@@ -1,11 +1,15 @@
...
-import store from '../redux/store';
+import store, { AppDispatch, AppState } from '../redux/store';
 import { multicallUpdateOutdatedListeners } from './multicall';
 import logger from 'logger';
 
 // -- Actions ---------------------------------------- //
-const updateMulticall = blockNumber => async (dispatch, getState) => {
+const updateMulticall = (blockNumber: number) => async (
+  dispatch: AppDispatch,
+  getState: () => AppState
```
Comment (jinchung) [inline src/redux/web3listener.ts:11]: `AppGetState` usage in this file as well

File: src/redux/web3listener.ts [hunk 7]
```diff
@@ -22,11 +26,17 @@ const debouncedUpdateMulticallListeners = debounce(blockNumber => {
   store.dispatch(updateMulticall(blockNumber));
 }, 1000);
 
-export const web3ListenerInit = () => {
-  web3Provider.pollingInterval = 10000;
-  web3Provider.on('block', debouncedUpdateMulticallListeners);
+let typedWeb3Provider = web3Provider as JsonRpcProvider | null;
```
[the remark this answers is not in the document]
Comment (jinchung) [reply src/redux/web3listener.ts:29]: ah ideally `handlers/web3` should be TS'd. I realize now that a lot of the items in `handlers` will be used even in the `redux` files so it may be beneficial to include these as part of your conversion

File: src/redux/userLists.ts [hunk 8]
```diff
@@ -21,16 +23,90 @@ const USER_LISTS_UPDATE_LISTS = 'userLists/USER_LISTS_UPDATE_LISTS';
...
 
   dispatch({ type: USER_LISTS_LOAD_REQUEST });
   try {
-    const defaultLists = DefaultTokenLists[network] || [];
-    const userLists = await getUserLists(network);
+    const defaultLists =
+      (DefaultTokenLists as Record<
+        Network,
+        typeof DefaultTokenLists[keyof typeof DefaultTokenLists]
+      >)[network] || [];
```
[the remark this answers is not in the document]
Comment (jinchung) [reply src/redux/userLists.ts:106]: this is a good point - `Network` is actually a bit of a confusing type - we need to actually split this out into 3 different types used for different purposes; can discuss further later

[the remark this answers is not in the document]
Comment (jinchung) [reply src/redux/userLists.ts:106]: in the interim, can this type be exposed in `references` itself where `DefaultTokenLists` comes from?

[the remark this answers is not in the document]
Comment (marco) [reply src/redux/userLists.ts]: Okay, for now I moved this type into `references` and now import it when doing the cast, and in the future we can split up `Network` like you referenced. Thanks!

File: src/redux/web3listener.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (marco) [reply src/redux/web3listener.ts:29]: I can start working on `handlers/web3` tomorrow morning!

File: src/entities/userLists.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+/**
+ * A single user list, with fields based on the data found in
+ * `src/references/default-token-lists.json`.
+ */
+export interface UserList {
```
Comment (jinchung) [inline src/entities/userLists.ts:5]: we have a pattern of exporting to `entities/index` (you can check out the other ones that are already in there) and only importing from `entities` instead of having to do `entities/userList` etc!

Comment (marco) [reply src/entities/userLists.ts:5]: Great, thanks! I pushed an update that should address this and your second comment. I'll also add the `handlers/web3` updates when I'm done with them later today.

File: src/redux/web3listener.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (marco) [reply src/redux/web3listener.ts:29]: @jinchung `handlers/web3` ended up being relatively involved, so I opened #2511 to track that separately. Thanks!

File: src/redux/settings.ts [hunk 10]
```diff
@@ -25,7 +26,51 @@ const SETTINGS_UPDATE_NETWORK_SUCCESS =
...
+    network: SettingsState['network'];
+  };
+}
+
+interface SettingsStateUpdateLanguageSuccessAction {
+  type: typeof SETTINGS_UPDATE_LANGUAGE_SUCCESS;
+  payload: SettingsState['language'];
+}
+
+export const settingsLoadState = () => async (dispatch: AppDispatch) => {
```
Comment (markdalgleish) [inline src/redux/settings.ts:73]: Same issue I noticed here: https://github.com/rainbow-me/rainbow/pull/2505#discussion_r743664238

tl;dr `AppDispatch` is being resolved as `AppDispatch<any>`, so it might be worth explicitly typing the actions here with `Dispatch<SettingsStateUpdateAction>`.

The tricky part is that some actions being dispatched in this file are coming from JS files so you might want to annotate those dispatches with `// @ts-expect-error` comments — that way you'll be prompted to come back and clean up this file when those files are converted to TS.

**Comment by marco** [comment]
Rebased and updated this one now that the the `web3` changes have been merged.

Proof of work:

Screen Shot 2021-12-07 at 11 20 39 AM

File: src/redux/web3listener.ts [hunk 11]
```diff
@@ -1,11 +1,15 @@
...
+import { ThunkDispatch } from 'redux-thunk';
+import { web3Provider } from '@rainbow-me/handlers/web3';
+import { multicallUpdateOutdatedListeners } from '@rainbow-me/redux/multicall';
+import store, { AppGetState, AppState } from '@rainbow-me/redux/store';
 import logger from 'logger';
 
 // -- Actions ---------------------------------------- //
-const updateMulticall = blockNumber => async (dispatch, getState) => {
+const updateMulticall = (blockNumber: number) => async (
+  dispatch: ThunkDispatch<AppState, unknown, never>,
```
Comment (jinchung) [inline src/redux/web3listener.ts:10]: why ThunkDispatch instead of AppDispatch ?

File: src/redux/userLists.ts [hunk 12]
```diff
@@ -82,9 +160,11 @@ export const userListsClearList = listId => (dispatch, getState) => {
...
-  const newList = { ...allNewLists[listIndex] };
-  newList.tokens = [];
-  allNewLists[listIndex] = newList;
+  if (listIndex) {
+    const newList = { ...allNewLists[listIndex] };
+    newList.tokens = [];
+    allNewLists[listIndex] = newList;
+  }
 
   dispatch({
```
Comment (jinchung) [inline src/redux/userLists.ts:169]: we can move this dispatch and line 173 (saveUserLists) into the if statement above

File: src/redux/userLists.ts [hunk 13]
```diff
@@ -114,19 +198,21 @@ export const userListsUpdateList = (assetAddress, listId, add = true) => (
...
-    const newList = { ...allNewLists[listIndex] };
-    newList.tokens = updatedListTokens;
-    allNewLists[listIndex] = newList;
+      // update the list
+      const newList = { ...allNewLists[listIndex] };
+      newList.tokens = updatedListTokens;
+      allNewLists[listIndex] = newList;
+    }
 
     dispatch({
```
Comment (jinchung) [inline src/redux/userLists.ts:217]: same thing we can move dispatch and saveUserLists into the if as well

File: src/redux/web3listener.ts (hunk 11 above)
Comment (jinchung) [reply src/redux/web3listener.ts:10]: same for the other redux files

Comment (marco) [reply src/redux/web3listener.ts:10]: This relates to https://github.com/rainbow-me/rainbow/pull/2498#discussion_r744327085, since there are other JS reducers that get included, the standard `AppDispatch` actually doesn't end up doing any type checking. Basically it resolves to `Dispatch<any>`. In this case I think `ThunkDispatch` instead of `Dispatch` because `multicallUpdateOutdatedListeners(blockNumber)` is being passed in instead of an object. For most of the other ones though, I used a `Dispatch` and specified which the valid actions are in order to get the type checking.

File: src/redux/userLists.ts (hunk 12 above)
Comment (marco) [reply src/redux/userLists.ts]: Okay, I updated this dispatch as well as the one below to be in the `if` statements, thanks!

File: src/redux/userLists.ts [hunk 14]
```diff
@@ -114,44 +198,46 @@ export const userListsUpdateList = (assetAddress, listId, add = true) => (
...
     // add or remove
-    const updatedListTokens = add
-      ? uniq(concat(allNewLists[listIndex].tokens, assetAddress))
-      : isArray(assetAddress)
-      ? without(allNewLists[listIndex].tokens, ...assetAddress)
-      : without(allNewLists[listIndex].tokens, assetAddress);
-    if (add) {
-      dispatch(emitAssetRequest(assetAddress));
-    }
+    if (listIndex) {
```
Comment (markdalgleish) [inline src/redux/userLists.ts:201]: `listIndex` could be `0`, right? Should this be `if (listIndex !== null)`?

File: src/redux/userLists.ts [hunk 15]
```diff
@@ -82,20 +160,26 @@ export const userListsClearList = listId => (dispatch, getState) => {
   });
 
   // update the list
-  const newList = { ...allNewLists[listIndex] };
-  newList.tokens = [];
-  allNewLists[listIndex] = newList;
+  if (listIndex) {
```
Comment (markdalgleish) [inline src/redux/userLists.ts:163]: Same here: `if (listIndex !== null)`?

File: src/redux/userLists.ts (hunk 14 above)
Comment (jinchung) [reply src/redux/userLists.ts:201]: ah yes very good point - nice catch!

Comment (marco) [reply src/redux/userLists.ts]: Yes definitely, I updated that as well as the one below, thanks!

**Review by jinchung — APPROVED** [review]
🥪

**Review by markdalgleish — APPROVED** [review]
🎉
